### DISCUSSION TITLE
TX16S MK2 turning off without warning

### DISCUSSION THREAD

**Discussion by pfeerick** [EdgeTX/edgetx#3666] [Q&A]
@tipiouk wrote:

> I bought my new TX16S mark2 two months ago. Since the beginning my radio sometimes turns off without warning even though the battery is fully charged and there are no bad contact. This hapened 6 times. I have to turn on the radio to use it. Have you ever heard of this problem? Do I have to update the software? It runs on Edge tx 2.6.0
> Sorry foer my poor english I'm french ;-)

**Comment by pfeerick** [comment]
So, a couple of questions :
1) How are you powering the radio - what battery pack/type? 
2) What exactly does the radio do when it turns off? Does it enter emergency mode on its own? Does it enter emergency mode when you try to power it back on?
3) Have you noticed any pattern as to when it turns off? Is it after a period of time, or perhaps when using the radio menus?

I've not heard of this sort of problem before, but there's always a first time.

**Reply by Doogster-hub** [reply]
I also have this problem with my tx16s, radio shut down with no warning at all while I was flying a very expensive aircraft leading to a crash due to having no radio link, plane is a total loss. transmitter went totally blank with no warnings whatsoever. When transmitter powered back on via power button EMERGENCY MODE was displayed. Battery and connections all good and secure. Did manage to get it to power up but froze on top page. Have since found the supplied 256 sd card is full. Deleted all sd logs and retried radio started as normal. Have not re bound to a model yet until I replace the sd card. I did not receive any sd card full warnings which would have avoided the crash and loss of aircraft, not to mention the risk of serious injury to spectators !! Is this an over sight by Radiomaster or a defect, if so it has the potential to cause serious damage to equipment and potentially people near by due to a simple problem.

**Comment by zyren** [comment]
Je suis désolé， je ne parle pas non plus français.J'utilise ChatGPT pour traduire ce passage， j'espère que vous pourrez comprendre.
Vous pouvez essayer de secouer vigoureusement le TX16S pour voir si ce problème se produit. Vous pouvez également essayer de mettre à jour le firmware vers la version 2.7.1 ou 2.8.4, puis effectuer un nouveau test pour voir si ce problème persiste.

**Comment by mha1** [comment]
Coincidentally I ran into the same SD card full situation today by triggering a screenshot which finally led to an EM (without restoring control), destroyed radio.yaml, an empty radio_error.yml and a non booting radio. Deleting some logs and restoring radio.yml from a backup got the radio working again.

I think we urgently need a warning page as early as in 2.8.5 similar to no failsafe or switch warnings at startup to inform the user if the SD card usage exceeds a certain safety margin (how about 50MB ) discouraging the user to fly until there is enough spare SD storage capacity. A better but properly tested solution that doesn't end in a totally unrecoverable situation should follow.

**Comment by Doogster-hub** [comment]
Exactly what happened to my radio, replaced sd card with larger capacity 32gb and restored with back up radio now working. Lost a model due to this easily avoidable problem. Taranis running open tx gave pending warning of sd card full. This would have avoided a lot of pain,expense and time

**Comment by mha1** [comment]
@pfeerick SD Card full is a situation that leads into a catastrophic event because the startup after the EM runs into the same problem. Maybe there are more scenarios (hopefully not) but I have tried to address the SD card full scenario.

I have added warnings similar to throttle warning at startup (opentx.cpp/checkAll()) as a quick measure to at least make users aware of a present or upcoming major problem:
- one for SD card sanity, maybe not required. If SDCARD is defined but not mounted at the time of the startup check a warning is issued. This really means something is totally f'd up, so idk.
- one for SD low storage. If SDCARD is defined and mounted and free storage is less than a configurable amount of MBytes a warning is issued. Given the common size of SD Cards being in the GBytes range I think reserving 50 to 100 MB to sustain a day of flying so wouldn't hurt. 

Of course both are warnings and can be ignored but ignoring them has the potential for a catastrophic event.

I think the warnings are mandatory but not enough. EdgeTX should be hardened against low SD storage events, i.e. defining an absolute minimum SD card free memory threshold, monitoring SD card free memory and taking appropriate action once the threshold is violated. Like shutting down logging, shutting down yaml writing, no more screenshots etc.

Please advise if starting with warnings is the right thing to do and if I you want me to proceed with a PR.

![image](https://github.com/EdgeTX/edgetx/assets/5615068/71e15773-9453-49b8-8612-7e4f53eb6f34)
![image](https://github.com/EdgeTX/edgetx/assets/5615068/39326846-bff4-4d9d-8ad7-3988b4c87908)

**Reply by pfeerick** [reply]
The warnings as suggested sound perfect in the first instance. The first one (not mounted) may not be needed, as there is already a No SD card warning (unless that doesn't trigger if say the SD card was corrupt / wrong FS format). 

Yes, logging and screenshots should be disabled/blocked at say 50MB of free space remaining, nothing else should need to be stopped as they are the consumers. I guess it could be as simple as a check when they are triggered, and if free space is insufficient, do nothing / hopefully trigger some sort of UI prompt without crashing the radio as this could be triggered at any time, and multiple consecutive times.

**Reply by mha1** [reply]
https://github.com/EdgeTX/edgetx/pull/3730

**Comment by Doogster-hub** [comment]
Hi AllYes absolutely agree, 50mb sounds about right as a safe buffer to trigger a warning as well👍RegardsDouglas MooreSent from my iPhoneOn 28 Jun 2023, at 08:15, Peter Feerick ***@***.***> wrote:﻿
The warnings as suggested sound perfect in the first instance. The first one (not mounted) may not be needed, as there is already a No SD card warning (unless that doesn't trigger if say the SD card was corrupt / wrong FS format).
Yes, logging and screenshots should be disabled/blocked at say 50MB of free space remaining, nothing else should need to be stopped as they are the consumers.
